prboom2/src/p_pspr.c

```diff
@@ -365,6 +424,36 @@ dboolean P_CheckAmmo(player_t *player)
   return false;
 }
 
+//
+// P_SubtractAmmo
+// Subtracts ammo, w/compat handling. In mbf21, use
+// readyweapon's "ammopershot" field if it's explicitly
+// defined in dehacked; otherwise use the amount specified
+// by the codepointer instead (Doom behavior)
+//
+// [XA] NOTE: this function is for handling Doom's native
+// attack pointers; of note, the new A_ConsumeAmmo mbf21
+// codepointer does NOT call this function, since it doesn't
+// have to worry about any compatibility shenanigans.
+//
+
+void P_SubtractAmmo(struct player_s *player, int compat_amt)
+{
+  int amount;
+  ammotype_t ammotype = weaponinfo[player->readyweapon].ammo;
+
+  if (mbf21 && ammotype == am_noammo)
+    return; // [XA] hmm... I guess vanilla/boom will go out of bounds then?
+
+  if (mbf21 && (weaponinfo[player->readyweapon].intflags & WIF_ENABLEAPS))
+    amount = weaponinfo[player->readyweapon].ammopershot;
+  else
+    amount = compat_amt;
+
+  if (!mbf21 || player->ammo[ammotype] >= amount)
```

Comment:
Not sure I understand this. The implication is somewhere in the code we called this method despite not having enough ammo? Is it possible? At any rate we should subtract the ammo we do have left in that case, and set the ammo to zero instead of skipping the subtraction.

Reply:
Yup, it's very easy to accidentally go into the negatives in vanilla DEHACKED -- random example, if you have a dual pistol weapon that calls A_FirePistol twice, and fire the weapon with 1 ammo remaining, you'll go to -1 ammo and the HUD will look bad. :P

Of course, preventing this kinda thing is exactly what the new `ammopershot` field is for, so this problem is avoidable simply by using mbf21 dehacked features correctly, but negative ammo is never a good thing in any case so I figured I'd guard against it outright.

This is really just a "nice to have", so if you feel strongly about it I can take it out -- or if you'd rather see the logic be "subtract, then set to zero if negative", that works for me -- bit cleaner that way, thinking on it :P

Reply:
Ya - just substract it and then do "if negative and mbf21, = 0" I think. Then at least if there's some weird mixup, it won't make the same mistake repeatedly (i.e., 1 will drop to 0 and then it should detect there's no ammo, hopefully).

Reply:
I took care of it 🎵